addons/pos_loyalty/static/src/js/ControlButtons/eWalletButton.js

```diff
@@ -0,0 +1,99 @@
+/** @odoo-module **/
+
+import PosComponent from 'point_of_sale.PosComponent';
+import ProductScreen from 'point_of_sale.ProductScreen';
+import Registries from 'point_of_sale.Registries';
+
+export class eWalletButton extends PosComponent {
+    getEWalletRewards(order) {
+        const claimableRewards = order.getClaimableRewards();
+        const eWalletRewards = claimableRewards.filter(({ reward }) => reward.program_id.program_type == 'ewallet');
+        return eWalletRewards;
+    }
+    getEWalletPrograms() {
+        return this.env.pos.programs.filter((p) => p.program_type == 'ewallet');
+    }
+    async onClickWalletButton() {
+        const order = this.env.pos.get_order();
+        const eWalletPrograms = this.env.pos.programs.filter((p) => p.program_type == 'ewallet');
+        const orderTotal = order.get_total_with_tax();
+        const eWalletRewards = this.getEWalletRewards(order);
+        if (orderTotal < 0 && eWalletPrograms.length >= 1) {
+            const eWalletPrograms = this.env.pos.programs.filter((p) => p.program_type == 'ewallet');
+            let selectedProgram = null;
+            if (eWalletPrograms.length == 1) {
+                selectedProgram = eWalletPrograms[0];
+            } else {
+                const { confirmed, payload } = await this.showPopup('SelectionPopup', {
+                    title: this.env._t('Refund with eWallet'),
+                    list: eWalletPrograms.map((program) => ({
+                        id: program.id,
+                        item: program,
+                        label: program.name,
+                    })),
+                });
+                if (confirmed) {
+                    selectedProgram = payload;
+                }
+            }
+            if (selectedProgram) {
+                const eWalletProduct = this.env.pos.db.get_product_by_id(selectedProgram.trigger_product_ids[0]);
+                order.add_product(eWalletProduct, {
+                    price: -orderTotal,
+                    merge: false,
+                    eWalletGiftCardProgram: selectedProgram,
+                });
+            }
+        } else if (eWalletRewards.length >= 1) {
```

Comment:
I'm wondering, shouldn't we still show all the ewallet programs with their balance even if there's nothing? Since the point of handling multiple ewallet (although the specs don't cover all the possible use cases) is also to be able to have different "domain" for each ewallet. Let's say for example you have one for everything and one for ecological products, the first one is empty and the second one has some money. With your implementation when clicking on the ewallet pay, it will uses the second one without asking the cashier (yes he'll be able to notice it by reading the small active program in the order summary and remove it but still it'll make them do more mistake).

Reply:
David asked me to implement this. I like your idea though.